src/lib/TelemetryClient.ts

```diff
@@ -0,0 +1,166 @@
+
+'use strict';
+
+/* tslint:disable:no-var-requires */
+const appInsights = require('applicationinsights-js').AppInsights;
+const glimpse = require('glimpse');
+const metadataRepository = require('../shell/repository/metadata-repository');
+const request = require('superagent');
+/* tslint:enable:no-var-requires */
+
+import moment = require('moment');
+import * as uuid from 'node-uuid';
+
+interface ITelemetryEvent {
+    name: string;
+    properties: { [key: string]: string; }; // map string->string
+    measurements: { [key: string]: number; }; // map string->number
+}
+
+class TelemetryClient {
+
+    private lastEventTime: moment.Moment;
+    private sessionId: string;
+    private telemetryConfig: Object;
+    private isTelemetryEnabled = true;
+
+    // we'll queue telemetry events until the telemetry config is downloaded and app insights is configured.  
+    private eventQueue: ITelemetryEvent[] = [];
+
+    constructor() {
+        this.sessionId = uuid.v4();
+        const self = this;
+        metadataRepository.registerListener((metadata) => {
+
+            if (!metadata.resources || !metadata.resources['telemetry-config']) {
+                self.isTelemetryEnabled = false;
+            }
+            else {
+                const uri = metadata.resources['telemetry-config'];
```

Comment:
It's only one extra request, so it isn't like it's repeated.  That said, I don't have strong opinions about it, other than what I have is implemented & seems to be working.